File: 2022/src/main14.ts
import fs from "fs";

function runSim(grid: Set<string>, maxY: number, floor: boolean) {
    let done = false;
    let count = 0;
    while (!done) {
        let x = 500;
        let y = 0;

        if (grid.has(`${x},${y}`)) {
            break;
        }

        let settled = false;
        while (!settled) {
            if (!grid.has(`${x},${y+1}`)) {
                y++;
            } else if (!grid.has(`${x-1},${y+1}`)) {
                x--;
                y++;
            } else if (!grid.has(`${x+1},${y+1}`)) {
                x++;
                y++;
            } else {
                settled = true;
                grid.add(`${x},${y}`);
                count++;
            }

            if (y > maxY) {
                settled = true;
                if (floor) {
                    grid.add(`${x},${y}`);
                    count++;
                } else {
                    done = true;
                }
            }
        }
    }

    return count;
}

function main() {
    const file = fs.readFileSync("14-input.txt").toString();

    const lines = file.split("\n");

    let grid = new Set<string>();
    for (const line of lines) {
        const points = line.split(" -> ");

        let from = points.shift();
        let to = points.shift();

        while (from && to) {
            let [fx,fy] = from.split(",").map(Number);
            let [tx,ty] = to.split(",").map(Number);

            for (let x = fx, y = fy; x !== tx || y !== ty; x += Math.sign(tx - x), y += Math.sign(ty - y)) {
                grid.add(`${x},${y}`);
            }
            grid.add(`${tx},${ty}`);

            from = to;
            to = points.shift();
        }
    }

    let maxY = -Infinity;
    for (const p of grid) {
        const [_,y] = p.split(",").map(Number);
        maxY = Math.max(maxY, y);
    }

    console.log(runSim(new Set(grid), maxY, false));
    console.log(runSim(new Set(grid), maxY, true));
}

main();

File: 2022/app/Main14.hs
module Main where

import Data.Char
import Data.Sort
import Debug.Trace
import qualified Data.Map as M
import qualified Data.Set as S
import Control.Monad
import Data.List.Split

data Point = Point Int Int deriving (Show, Eq, Ord)

parsePair :: String -> Point
parsePair s = let [x,y] = map read . splitOn "," $ s in Point x y

parse :: String -> [[Point]]
parse = (map (map parsePair . (splitOn " -> "))) . lines

lerp :: Point -> Point -> [Point]
lerp (Point x1 y1) (Point x2 y2) = let
    dx = x2 - x1
    dy = y2 - y1
    steps = max (abs dx) (abs dy)
    in map (\i -> Point (x1 + (dx * i) `div` steps) (y1 + (dy * i) `div` steps)) [0..steps]

lerpPairs :: [Point] -> [Point]
lerpPairs (x:y:xs) = lerp x y ++ lerpPairs (y:xs)
lerpPairs _ = []

settle :: S.Set Point -> Int -> Bool -> Point -> (S.Set Point, Int)
settle walls maxY floor p =
  let (Point x y) = p
      down = Point x (y + 1)
      left = Point (x - 1) (y+1)
      right = Point (x + 1) (y+1)
  in
    if (S.member p walls) then (walls, 0)
    else if y > maxY then
      if floor then (S.insert p walls, 1)
      else (walls, 0)
    else if not (S.member down walls) then settle walls maxY floor down
    else if not (S.member left walls) then settle walls maxY floor left
    else if not (S.member right walls) then settle walls maxY floor right
    else (S.insert p walls, 1)

go :: S.Set Point -> Int -> Bool -> Int -> (S.Set Point, Int)
go walls maxY floor count =
  let (walls', c) = settle walls maxY floor (Point 500 0)
  in if c == 0 then (walls', count) else go walls' maxY floor (count+1)

runSim :: S.Set Point -> Bool -> Int
runSim walls floor =
  let maxY = maximum . map (\(Point _ y) -> y) . S.toList $ walls
      (_, count) = go walls maxY floor 0
  in count

main :: IO ()
main =
  readFile "14-input.txt" >>= \content ->

--  (mapM_ putStrLn (lines content)) >>

  let walls = S.fromList . concatMap lerpPairs $ parse content
  in return () >>

  print (runSim walls False) >>
  print (runSim walls True) >>

  return ()
